x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -240,35 +241,154 @@ public Automaton allowedActionsMatcher(String index) {
     }
 
     /**
-     * Authorizes the provided action against the provided indices, given the current cluster metadata
+     * Represents the set of data required about an IndexAbstraction (index/alias/datastream) in order to perform authorization on that
+     * object (including setting up the necessary data structures for Field and Document Level Security).
      */
-    public Map<String, IndicesAccessControl.IndexAccessControl> authorize(String action, Set<String> requestedIndicesOrAliases,
-                                                                          Map<String, IndexAbstraction> lookup,
-                                                                          FieldPermissionsCache fieldPermissionsCache) {
-        // now... every index that is associated with the request, must be granted
-        // by at least one indices permission group
-        Map<String, Set<FieldPermissions>> fieldPermissionsByIndex = new HashMap<>();
-        Map<String, DocumentLevelPermissions> roleQueriesByIndex = new HashMap<>();
-        Map<String, Boolean> grantedBuilder = new HashMap<>();
+    private static class IndexResource {
+        /**
+         * The name of the IndexAbstraction on which authorization is being performed
+         */
+        private final String name;
+
+        /**
+         * The type of the IndexAbstraction on which authorization is being performed
+         */
+        private final IndexAbstraction.Type type;
+
+        /**
+         * The names of the concrete indices to which the IndexAbstraction refers.
+         * If {@link #type} is {@link IndexAbstraction.Type#CONCRETE_INDEX}, this will be a singleton set of {@link #name}.
+         * For all other types, it will be the backing indices for the named object.
+         */
+        private final Collection<String> concreteIndices;
+
+        /**
+         * The name of the parent datastream, if any - otherwise {@code null}.
+         */
+        @Nullable
+        private final String parentDataStream;
+
+        private IndexResource(String name, IndexAbstraction.Type type, Collection<String> concreteIndices, String parentDataStream) {
+            if (Assertions.ENABLED) {
+                assert name != null : "Resource name cannot be null";
+                if (type == IndexAbstraction.Type.CONCRETE_INDEX) {
+                    assert concreteIndices.size() <= 1 : "An object of type " + type + " cannot have multiple indices";
+                    assert concreteIndices.isEmpty() || concreteIndices.contains(name) : "An object of type "
+                        + type
+                        + " must reference itself";
```

Comment:
Why do we allow both an empty collection and a singleton collection for concrete Index type? I'd prefer to pick either one of them. IIUC, the subtle difference between empty and singleton is that an empty collection indicates the concrete index does not exist in the cluster state yet. But do we care this difference for authorization?

Reply:
Because I was trying to avoid implementing behavioural changes, and that's what the existing code does - a missing index is treated like a concrete index with no expanded names.